tool/src/org/antlr/v4/tool/GrammarTransformPipeline.java

```diff
@@ -509,4 +517,43 @@ public GrammarRootAST extractImplicitLexer(Grammar combinedGrammar) {
 		return lexerAST;
 	}
 
+	public void expandActionTemplates(GrammarRootAST root) {
+		String fileName = root.g.getActionTemplates();
+		try {
+			STGroupFile actionTemplates = new STGroupFile(fileName);
+			TreeVisitor v = new TreeVisitor(new GrammarASTAdaptor());
+			v.visit(root, new TreeVisitorAction() {
+				@Override
+				public Object pre(Object t) {
+					if (((GrammarAST) t).getType() == ANTLRParser.ACTION) {
+						return expandActionTemplate((GrammarAST) t, actionTemplates);
+					}
+					return t;
+				}
+				@Override
+				public Object post(Object t) {
+					return t;
+				}
+			});
+		} catch (IllegalArgumentException e) {
+			if (e.getMessage() != null && e.getMessage().startsWith("No such group file")) {
+				tool.errMgr.toolError(ErrorType.CANNOT_FIND_ACTIONS_TEMPLATE_FILE_GIVEN_ON_CMDLINE, e, fileName, root.g.name);
+			} else {
+				tool.errMgr.toolError(ErrorType.ERROR_READING_ACTION_TEMPLATES_FILE, e, fileName, e.getMessage());
+			}
+		} catch (STException e) {
+			tool.errMgr.toolError(ErrorType.ERROR_READING_ACTION_TEMPLATES_FILE, e, fileName, e.getMessage());
+		}
+	}
+
+	public GrammarAST expandActionTemplate(GrammarAST t, STGroupFile actionTemplates) {
+		String grammarFileInfo = t.g.fileName + " " + t.getLine() + 1 + ":" + t.getCharPositionInLine();
```

Comment:
This is absolute rubbish - would love your ideas about how to connect up the error reporting from StringTemplate to the line/column offsets in the ANTLR grammar.

Reply:
I'm investigating using STErrorListener to connect StringTemplate's error reporting to ANTLR's here